File: cpp_solutions/552.cpp
typedef long long LL;
class Solution {
public:
    int checkRecord(int n) {
        LL mod = 1e9 + 7;
        vector<vector<LL>> dp(3, vector<LL>(2, 0));
        vector<vector<LL>> dpTemp(3, vector<LL>(2, 0));
        dp[0][0] = 1; //P
        dp[1][0] = 1; //L
        dp[0][1] = 1; //A
        
        for (int i = 1; i < n; ++i) {
            dpTemp[0][0] = (dp[0][0] + dp[1][0] + dp[2][0]) % mod;
            dpTemp[1][0] = dp[0][0] % mod;
            dpTemp[2][0] = dp[1][0] % mod;
            
            dpTemp[0][1] = (dpTemp[0][0] + dp[0][1] + dp[1][1]  + dp[2][1]) % mod;
            dpTemp[1][1] = dp[0][1] % mod;
            dpTemp[2][1] = dp[1][1] % mod;
            dp = dpTemp;
        }
        
        LL res = 0;
        for (int j = 0; j < 3; ++j) {
            for (int k = 0; k < 2; ++k) {
                res += dp[j][k];
                res %= mod;
            } 
        }
        
        return res;
    }
};

File: rust_solutions/552.rs
impl Solution {
    pub fn check_record(n: i32) -> i32 {
        let MOD = 1000000007usize;
        let mut dp: Vec<usize> = vec![1, 1, 0, 1, 0, 0];
        let mut temp_dp = dp.clone();

        for i in 1..n {
            temp_dp[0] = (dp[0] + dp[1] + dp[2]) % MOD;
            temp_dp[1] = dp[0];
            temp_dp[2] = dp[1];
            temp_dp[3] = dp.iter().sum::<usize>() % MOD;
            temp_dp[4] = dp[3];
            temp_dp[5] = dp[4];
            std::mem::swap(&mut dp, &mut temp_dp);
        }

        (dp.iter().sum::<usize>() % MOD) as i32
    }
}
